Fix `onFocusOutside` callback on unmount

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

Alternative to https://github.com/WordPress/gutenberg/pull/71252


## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->
Closes https://github.com/WordPress/gutenberg/issues/71281

<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

When focus is moved outside an element that has the `useFocusOutside` hook, the callback should run, even if it's removed due to unmounting.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The current `trunk` implementation of `useFocusOutside` works by:

- Attaching `onFocus` and `onBlur` listeners to the container
- `onBlur` will run anytime focus is moved and queue a cancellable timeout that will run after focus has resolved.
- `onFocus` runs before the timeout settles. If the container runs this `onFocus` check, then focus is resolving within the container. This focus event cancels the onFocusOutside callback from the blur event.
- If no focus event happens within the container, then the blur event runs.

This PR tries removing the onFocus check and setTimeout cancellation system with one onBlur event. This onBlur event checks `event.currentTarget` for access to the container to check if focus is within the container. Additionally, if the container spawns another popover, the popover will remain open because it also checks to see if there is an element within it that is marked as `aria-expanded=true` and `aria-haspopover=true`

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

|Before|After|
|-|-|
|<!-- Before screenshot here -->|<!-- After screenshot here -->|

jeryj committed:
Container Ref + onBlur for useFocusOutside hook

Tries removing all the setTimeout and blur/focus checks with a simpler onBlur and container check for useFocusOutside

jeryj committed:
Fix broken build for with-focus-outside

jeryj committed:
Wrap onBlur check in 0 setTimeout to delay check til focus has resolved

jeryj committed:
Only wrap the callback in setTimeout to fix focustrap

jeryj commented on packages/compose/src/hooks/use-focus-outside/index.ts:
```diff
@@ -64,120 +13,49 @@ type UseFocusOutsideReturn = {
-			event.relatedTarget?.closest( ignoreForRelatedTarget )
-		) {
-			return;
-		}
```
We'd need to check if this is still needed. It's only used [in one place](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/inserter/media-tab/media-tab.js#L87)

jeryj committed:
Remove tests no longer needed due to implementation change

jeryj committed:
Catch cases where a popover spawns another popover

jeryj committed:
Use event.currentTarget instead of ref-based approach for container

jeryj committed:
Cleanup: remove added ref from use dialog

jeryj committed:
Cleanup: remove added line

jeryj committed:
Remove usage of data-unstable-ignore-focus-outside-for-relatedtarget

jeryj committed:
Fix failing rename block from dialog test

I can't tell why this failing. It works when I manually try it. Everything appears right, but the saveButton.click() throws a wrench it in somehow.

jeryj commented on test/e2e/specs/editor/various/block-renaming.spec.js:
```diff
@@ -79,7 +79,11 @@ test.describe( 'Block Renaming', () => {
 			await expect( saveButton ).toBeEnabled();
 
-			await saveButton.click();
+			await page.keyboard.press( 'Tab' );
+			await page.keyboard.press( 'Tab' );
+
+			await expect( saveButton ).toBeFocused();
+			await page.keyboard.press( 'Enter' );
```
This shouldn't be necessary to change. Something about `await saveButton.click();` was throwing in a wrench in it. If I try the previous test manually it works. I debugged for awhile but couldn't isolate the reason, so moved on using this test.

I have a concern that it's failing for a valid reason, but since I can't reproduce when I do the test manually, I think it's OK.

jeryj commented on packages/compose/src/hooks/use-focus-outside/test/index.js:
```diff
@@ -85,56 +95,4 @@ describe( 'useFocusOutside', () => {
 		expect( mockOnFocusOutside ).toHaveBeenCalled();
 	} );
-
-	it( 'should not call handler if focus shifts outside the component when the document does not have focus', async () => {
```
This test is an implementation detail on useFocusOutside which is no longer valid. We don't test for document.hasFocus()

jeryj commented on packages/compose/src/hooks/use-focus-outside/test/index.js:
```diff
@@ -85,56 +95,4 @@ describe( 'useFocusOutside', () => {
-		mockedDocumentHasFocus.mockRestore();
-	} );
-
-	it( 'should cancel check when unmounting while queued', async () => {
```
This is the behavior we're changing. We want the callback because focus is moved outside the component.

jeryj commented on packages/compose/src/hooks/use-focus-outside/index.ts:
```diff
@@ -3,58 +3,7 @@
-}
-
 type UseFocusOutsideReturn = {
-	onFocus: React.FocusEventHandler;
-	onMouseDown: React.MouseEventHandler;
-	onMouseUp: React.MouseEventHandler;
-	onTouchStart: React.TouchEventHandler;
-	onTouchEnd: React.TouchEventHandler;
```
These were all necessary to catch focus events that happened within the container in order to cancel the timeout. Since we're not cancelling a timeout on focus, we don't need any of these or the code that supports them.

jeryj commented on packages/components/src/higher-order/with-focus-outside/test/index.tsx:
```diff
@@ -92,32 +92,4 @@ describe( 'withFocusOutside', () => {
 		await waitFor( () => expect( onFocusOutside ).toHaveBeenCalled() );
 	} );
-
-	it( 'should not call handler if focus shifts outside the component when the document does not have focus', async () => {
```
This test is an implementation detail on useFocusOutside which is no longer valid. We don't test for document.hasFocus()

jeryj commented on packages/components/src/higher-order/with-focus-outside/test/index.tsx:
```diff
@@ -92,32 +92,4 @@ describe( 'withFocusOutside', () => {
-		await waitFor( () => expect( onFocusOutside ).not.toHaveBeenCalled() );
-	} );
-
-	it( 'should cancel check when unmounting while queued', async () => {
```
This is the behavior we're changing. We want the callback because focus is moved outside the component.

jeryj commented on packages/block-editor/src/components/inserter/media-tab/media-tab.js:
```diff
@@ -72,19 +72,11 @@ function MediaTab( {
 									} }
 									className="block-editor-inserter__media-library-button"
 									variant="secondary"
-									data-unstable-ignore-focus-outside-for-relatedtarget=".media-modal"
```
If this PR is merged, `data-unstable-ignore-focus-outside-for-relatedtarget` will not do anything, so we should remove it.

jeryj commented on packages/block-editor/src/components/inserter/media-tab/media-tab.js:
```diff
@@ -72,19 +72,11 @@ function MediaTab( {
-										// focus event and therefore a `blur` event is triggered on a different
-										// element, which doesn't contain the `data-unstable-ignore-focus-outside-for-relatedtarget`
-										// attribute making the Inserter dialog to close.
-										event.target.focus();
```
This was necessary to cancel the timeout when the blur event happened. It's no longer necessary. Also, the inserter isn't a dialog anymore, so it wouldn't be necessary anyways.

aduth commented on packages/compose/src/hooks/use-focus-outside/index.ts:
```diff
@@ -72,112 +21,48 @@ export default function useFocusOutside(
+			! containerElement ||
+			! containerElement.contains( relatedTarget )
+		) {
+			// Wrapping in setTimeout to run the check after focus and rendering resolves.
+			setTimeout( () => {
```
Are we not able to get rid of the `setTimeout` ? Sorry if this was explained somewhere, I didn't see it.

jeryj replied:
I was getting failed tests when I had it, but I think it was actually due to the component unmounting then remounting. So, removing the setTimeout and then fixing the bugs is the better approach. I think they will mostly be related to popups spawning other popups. And the fix will be to add the correct attributes (aria-haspopup and aria-expanded="true"). 

Pushed the change and am logging off for the day. We'll see how many things need fixed as a result 😅

jeryj replied:
I think running the callback on blur is the wrong approach. It should run after focus resolves, because we'll get into weird states where the onClose callback sets states before the onMouseUp event finishes. This means you can't click a button to open the popover, then click the same button to close it (the popover will reopen). I think the `trunk` approach of queueing and canceling timeouts is a really clever and correct approach to this. 

I opened https://github.com/WordPress/gutenberg/pull/71294/ on top of this branch to remove all the extra browser hacks that I don't think are necessary anymore. It would simplify the code if we can remove it all. I'm not 100% sure if we still support a browser that needs that code though.

jeryj committed:
Remove setTimeout and modify haspopup attribute check

The haspopup attribute check was too specific. It can be values other than popup, such as dialog. I also fixed an instance on the keyboard shortcut where it needed a haspopup attribute but wasn't present. I like this approach as it happens to enforce a11y best practices.

jeryj committed:
Add setTimeout and notes about bug
